fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/TopologicalSort.java

```diff
@@ -0,0 +1,716 @@
+                    // In that case we need to abandon the current level and search on the level above (making that
+                    // level the current level). If we reach level -1 (i.e., we reach the end of the iterator at level 0
+                    // we are done.
+                    // If we do find an element not violating any constraints on the current level we conceptually
+                    // bind it the element we found and continue on downward.
+                    //
+                    final boolean foundOnLevel = nextOnLevel(currentIterator);
+                    if (!foundOnLevel) {
+                        // back tracking -- need to clear out the current iterator
+                        iterators.set(currentLevel, null);
+                    }
+                    currentLevel = foundOnLevel
+                                   ? currentLevel + 1
+                                   : currentLevel - 1;
+
+                    if (currentLevel == -1) {
+                        return endOfData();
+                    }
+                } while (bound.size() < set.size()); // as long as we still have to find a binding
+
+                return iterators.stream()
+                        .map(PeekingIterator::peek)
+                        .collect(ImmutableList.toImmutableList());
+            }
+
+            @SuppressWarnings({"squid:S135", "UnstableApiUsage", "PMD.AvoidBranchingStatementAsLastInLoop"})
+            private boolean nextOnLevel(final PeekingIterator<T> currentIterator) {
+                while (currentIterator.hasNext()) {
+                    final T next = currentIterator.peek();
+
+                    // check if it is bound already; t is invisible to this loop if it is bound
+                    if (bound.contains(next)) {
+                        currentIterator.next();
+                        continue;
+                    }
+
+                    // this level can be bound now
+                    bound.add(next);
+
+                    final Set<T> targets = usedByMap.get(next);
+                    final ImmutableSet.Builder<T> newlyEligibleElementsBuilder = ImmutableSet.builderWithExpectedSize(targets.size());
+                    for (final T target : targets) {
+                        final int newInDegree = inDegreeMap.compute(target, (k, v) -> Objects.requireNonNull(v) - 1);
+                        Verify.verify(newInDegree >= 0);
+                        if (newInDegree == 0) {
+                            newlyEligibleElementsBuilder.add(target);
+                        }
+                    }
+
+                    if (bound.size() < set.size()) {
+                        newlyEligibleElementsBuilder
+                                .addAll(eligibleElementSets.get(bound.size() - 1))
+                                .build();
+
+                        //eligibleElementSets.set(bound.size(), Sets.union(eligibleElementSets.get(bound.size() - 1), newlyEligibleElements));
```

Comment:
Yes, as it turns out, creating sets that are views (unions) of other sets are a bad idea if they are the equivalent nesting dolls as your set then turns into a linked list and look ups are effectively linear instead of effectively constant.